acceptance/lib/pxp-agent/test_helper.rb

```diff
@@ -50,3 +51,99 @@ def kill_pcp_broker(host)
     on(host, "kill -9 #{pid}")
   end
 end
+
+# Make an rpc_blocking_request to pxp-agent
+# Reference: https://github.com/puppetlabs/pcp-specifications/blob/master/pxp/request_response.md
+# @param broker the hostname or beaker host object of the machine running PCP broker
+# @param targets array of PCP identities to send request to
+#                e.g. ["pcp://client01.example.com/agent","pcp://client02.example.com/agent"]
+# @param pxp_module which PXP module to call, default pxp-module-puppet
+# @param action which action in the PXP module to call, default run
+# @param params params to send to the module. e.g for pxp-module-puppet:
+#               {:env => [], :flags => ['--noop', '--onetime', '--no-daemonize']}
+# @return hash of responses, with target identities as keys, and value being the response message for that identity as a hash
+# @raise String indicating something went wrong, test case can use to fail
+def rpc_blocking_request(broker, targets,
+                         pxp_module = 'pxp-module-puppet', action = 'run',
+                         params)
+  # Event machine is required by the ruby-pcp-client gem
+  # https://github.com/puppetlabs/ruby-pcp-client
+  em = Thread.new { EM.run }
+
+  mutex = Mutex.new
+  have_response = ConditionVariable.new
+  responses = Hash.new
+
+  client = PCP::Client.new({
+    :server => broker_ws_uri(broker),
+    :ssl_cert => "../test-resources/ssl/certs/controller01.example.com.pem",
+    :ssl_key => "../test-resources/ssl/private_keys/controller01.example.com.pem"
+  })
+
+  client.on_message = proc do |message|
+    mutex.synchronize do
+      resp = {
+        :envelope => message.envelope,
+        :data     => JSON.load(message.data),
+        :debug    => JSON.load(message.debug)
+      }
+      responses[resp[:envelope][:sender]] = resp
+      have_response.signal
+    end
+  end
+
+  client.connect()
+  if !client.associated?
+    raise "Controller PCP client failed to associate with pcp-broker on #{broker}"
+  end
+
+  message = PCP::Message.new({
+    :message_type => 'http://puppetlabs.com/rpc_blocking_request',
+    :targets => targets
+  })
+
+  message.data = {
+    :transaction_id => SecureRandom.uuid,
+    :module         => pxp_module,
+    :action         => action,
+    :params         => params
+  }.to_json
+
+  message_expiry = 10 # Seconds for the PCP message to be considered failed
+  rpc_action_expiry = 60 # Seconds for the entire RPC action to be considered failed
+  message.expires(message_expiry)
+
+  client.send(message)
+
+  begin
+    Timeout::timeout(rpc_action_expiry) do
+      done = false
+      loop do
+        mutex.synchronize do
+          have_response.wait(mutex)
+          done = have_all_rpc_responses?(targets, responses)
+        end
+        break if done
+      end
+    end
+  rescue Timeout::Error
+    if !have_all_responses?
```

Comment:
You need to acquire the mutex to call check_all_responses() because, after the timeout expiration, you may have concurrent message handler that modifies the `responses` hash.